push/src/main/java/org/aerogear/mobile/push/PushService.java

```diff
@@ -0,0 +1,313 @@
+                                firebaseMessaging.subscribeToTopic(catgory);
+                            }
+                        }
+
+                        firebaseMessaging.subscribeToTopic(unifiedPushCredentials.getVariant());
+
+                        callback.onSuccess();
+                        break;
+                    default:
+                        callback.onError(new HttpException());
+                        break;
+                }
+            });
+
+        } catch (JSONException e) {
+            MobileCore.getLogger().error(e.getMessage(), e);
+            callback.onError(e);
+        }
+
+    }
+
+    public void unregisterDevice(final Callback callback) {
+        unregisterDevice(new UnifiedPushConfig(), callback);
+    }
+
+    public void unregisterDevice(final UnifiedPushConfig unifiedPushConfig,
+                    final Callback callback) {
+
+        String authHash = getHashedAuth(unifiedPushCredentials.getVariant(),
+                        unifiedPushCredentials.getSecret().toCharArray());
+
+        final HttpRequest httpRequest = core.getHttpLayer().newRequest();
+        httpRequest.addHeader("Authorization", authHash);
+        httpRequest.delete(url + registryDeviceEndpoint + "/"
+                        + FirebaseInstanceId.getInstance().getToken());
+
+        final HttpResponse httpResponse = httpRequest.execute();
+        httpResponse.onSuccess(() -> {
+            switch (httpResponse.getStatus()) {
+                case HTTP_NO_CONTENT:
+
+                    FirebaseMessaging firebaseMessaging = FirebaseMessaging.getInstance();
+
+                    for (String category : unifiedPushConfig.getCategories()) {
+                        firebaseMessaging.unsubscribeFromTopic(category);
+                    }
+
+                    firebaseMessaging.unsubscribeFromTopic(unifiedPushCredentials.getVariant());
+
+                    callback.onSuccess();
+                    break;
+                default:
+                    callback.onError(new HttpException());
+                    break;
+            }
+        });
+
+    }
+
+    private String getHashedAuth(String username, char[] password) {
+        StringBuilder headerValueBuilder = new StringBuilder("Basic").append(" ");
+        String unhashedCredentials = username + ":" + String.valueOf(password);
+        String hashedCrentials =
+                        Base64.encodeToString(unhashedCredentials.getBytes(), Base64.NO_WRAP);
+        return headerValueBuilder.append(hashedCrentials).toString();
+    }
+
+
+    /**
+     * When a push message is received, all main thread handlers will be notified on the main(UI)
+     * thread. This is very convenient for Activities and Fragments.
+     *
+     * @param handler a handler to added to the list of handlers to be notified.
+     */
+    public static void registerMainThreadHandler(MessageHandler handler) {
+        MAIN_THREAD_HANDLERS.add(handler);
```

Comment:
If 2 threads calls this method, we will get a ConcurrentModificationException. If threading is not an issue, ignore this, otherwise consider using a synchronised collection.

Reply:
Not sure if I got you here! `MAIN_THREAD_HANDLERS` is a _static_ `List` of `MessageHandler`

Reply:
Exactly. Of two different threads will call `registerMainThreadHandler` at the same time, we will get a `ConcurrentModificationException` here, unless MAIN_THREAD_HANDLERS is a synchronised List

Reply:
Ok, not I got your point, sorry!

Reply:
I can't be wrong but in theory I can't have a static synchronized list. Should I solve it doing:

```java
public static void unregisterBackgroundThreadHandler(final MessageHandler handler) {
    nonNull(handler, "handle");
    synchronized (PushService.class) {
        BACKGROUND_THREAD_HANDLERS.remove(handler);
    }
}
```

Reply:
I'm not sure I got you.
I think you could do:

```java
private static final List<MessageHandler> MAIN_THREAD_HANDLERS = Collections.synchronizedList(new ArrayList<>());
```

the same for the background handlers.

Reply:
Hrm, I was about to say I can't have `static synchronized` field. Nevermind!

Reply:
These are synchyronized already https://github.com/aerogear/aerogear-android-sdk/pull/151/files#diff-24684cb8332f25e88fee7344b4b376a7R46 don't double synchronize.